### PR TITLE
C bindings to facilitate static FFI bindings

### PR DESCRIPTION
**Pull request by geoffder**
As [discussed](https://github.com/elalish/manifold/discussions/284), I've written C bindings to facilitate bindings through FFI from static compiled languages. I've been writing bindings to Manifold from [OCaml](https://github.com/geoffder/OManifold) using these bindings to make sure that they work as intended. They aren't polished and fully documented yet, but the API is covered and integration with the garbage collector seems correct.

The changes mostly live under [bindings/c](https://github.com/geoffder/manifold/tree/cbind/bindings/c), but the Manifold build/source/API has been touched in a few places, namely:

- sub-libraries are compiled as `OBJECT` and the `fPIC` flag is used when building the `c` bindings to allow more painless relocation of the archives/objects that is required for statically bundling them into executables in binding host languages (ran into this building the OCaml bindings)
- GPU `ExecutionPolicy` is avoided when the sdf parameter to `LevelSet` is not a static lambda (results in runtime failure)
- Optional control of `ExecutionPolicy` has been added to `LevelSet` to allow languages with thread/interpreter locks to avoid parallel access to the sdf closure
- `Decompose` has been broken up to allow the `c` bindings to know how many `Manifold` allocations will be needed in advance without needing to repeat the expensive part of the method.

### CODE REVIEW

**Comment by elalish** [comment]
Thanks! Hopefully the CLA is no big deal? Any thoughts on what it'll take to get it passing our CI? This is great, but my main concern is it's a lot of code that'll need to be updated as other developments happen. What do you think is the best way to go about this? For instance, #290 is steadily growing in scope.

**Comment by geoffder** [comment]
Ah it appears `std::optional` which I used to allow optional control over `ExecutionPolicy` for `LevelSet` wasn't available until C++17, so that is causing some failures.

**Comment by elalish** [comment]
I don't have any problem with pushing us up to C++17. If it helps to update the Ubuntu version of our CI, that's fine too.

**Comment by pca006132** [comment]
I think we were already using C++17 somewhere in the code, but the public API is C++14, so changing the version should not require updating Ubuntu. I think it would be better if we can reduce the size of this C binding, I think it can be reduced considering most of them are merely converting the fields to the C equivalent one, perhaps we can do some template magic here. (haven't look at it in details yet)

**Comment by pca006132** [comment]
After reading this a bit more carefully, it seems that it is hard to simplify this by using templates. For the API design I guess we can discuss about it later when there are other users of this C binding? I am not using the binding so it is hard for me to think about the API design.

**Comment by geoffder** [comment]
> After reading this a bit more carefully, it seems that it is hard to simplify this by using templates. 

I was using a template for the array conversion helper initially for example, but then I rewrote without it since there were only a couple instantiations anyway.

> For the API design I guess we can discuss about it later when there are other users of this C binding? I am not using the binding so it is hard for me to think about the API design.

For an idea of how these are used you can look at my C stub generation [descriptions](https://github.com/geoffder/OManifold/blob/main/c/manifold_c_function_descriptions.ml), and one of the [modules](https://github.com/geoffder/OManifold/blob/main/lib/manifold.ml) using them. Each language will have it's own quirks in interfacing with C of course, but I think that a lot of this pattern is preserved.

```ocaml
type t = C.Types.Manifold.t Ctypes_static.ptr

let size = C.Funcs.manifold_size () |> size_to_int
let destruct t = C.Funcs.destruct_manifold t

let alloc () =
  let finalise = Mem.finaliser C.Types.Manifold.t destruct in
  let buf = Mem.allocate_buf ~finalise size in
  buf, Ctypes_static.(Ctypes.coerce (ptr void) (ptr C.Types.Manifold.t) buf)

let add a b =
  let buf, t = alloc () in
  let _ = C.Funcs.manifold_union buf a b in
  t
```

Taking pre-allocated memory as an argument makes the process of ensuring it is registered with the host language properly fairly simple.

For a bit more complicate of an example using the additional machinery around `Decompose` that I added:

```ocaml
let decompose t =
  let comps_buf, comps = Components.alloc () in
  let _ = C.Funcs.manifold_get_components comps_buf t in
  let len = size_to_int @@ C.Funcs.manifold_components_length comps in
  let bufs = Ctypes.(CArray.make (ptr void) len)
  and ts = ref [] in
  for i = 0 to len - 1 do
    let buf, man = alloc () in
    Ctypes.CArray.set bufs i buf;
    ts := man :: !ts
  done;
  let _ = C.Funcs.manifold_decompose (Ctypes.CArray.start bufs) t comps in
  !ts
```

For a non-ocaml version of following this kind of C API binding scheme, there are the [olm python bindings](https://gitlab.matrix.org/matrix-org/olm/-/blob/master/python/olm/account.py), which as I mentioned in the discussion, is the library I used as an example. Minus the object-orientation, I used these official python bindings as a guide when I wrote mine in [ocaml](https://github.com/geoffder/olm-ml).

More people working on binding through these from other languages would certainly help to catch any rough edges that remain. They had to get a handful of fixes as I worked through them myself, like switching all arrays to pointers (including the dbl ptrs for the opaque types).

**Comment by geoffder** [comment]
- After figuring out how to point emscripten to the correct overload of `Decompose`, the WASM bindings are building again. 
- There is one test failing on Mac and I'm not sure why, I haven't touched anything in Polygon
- Despite bumping ubuntu to 22.04, the update to using the variables in the new `material.h` (`pbrmaterial.h` deprecated in assimp) is still causing a breakage. The nix builds fine, so obviously ubuntu is still lagging. Maybe some template conditionals are in order for compatibility?

**Comment by elalish** [comment]
Nice work! The Assimp stuff is not core to library, so feel free to hack in whatever's necessary to make it function. 

The `Boolean.Close` failure on mac is indeed interesting. I agree that's unlikely to be your fault, though it might be nice to make a small PR with only the C++17 bump to see if it repros there. In any case, I don't think it's your fault. You can change it to `TEST(Boolean, Close_DISABLED)` to get the CI passing. That'll remind me to look into it.

**Comment by geoffder** [comment]
I tried adding the `_DISABLED` suffix, but it still runs and fails on the test, so I probably haven't executed on your suggestion properly. 

I opened https://github.com/elalish/manifold/pull/293 to test whether the std version bump is responsible, but it passed.

File: test/mesh_test.cpp
Comment (elalish) [inline test/mesh_test.cpp:1057]: Sorry, meant `DISABLED_Close` - I always get that wrong...

File: test/meshIO/src/meshIO.cpp
Comment (elalish) [inline test/meshIO/src/meshIO.cpp:30]: 👍

File: bindings/c/box.cpp
Comment (elalish) [inline bindings/c/box.cpp:14]: What does this syntax mean? Is it just constructing a box at the pointer location `mem` and assuming there's enough space allocated there?

File: bindings/c/include/conv.hh
Comment (elalish) [inline bindings/c/include/conv.hh:10]: What is this for?

File: bindings/c/include/conv.hh
Comment (elalish) [inline bindings/c/include/conv.hh:1]: Why the `.hh` extension?

File: bindings/c/include/types.h
Comment (elalish) [inline bindings/c/include/types.h:30]: I take it this is because `glm` isn't a c library?

File: bindings/c/meshio.cpp
Comment (elalish) [inline bindings/c/meshio.cpp:1]: Do you even want to make bindings for this? MeshIO is nothing special, just a testing convenience.

File: bindings/c/testc.c
Comment (elalish) [inline bindings/c/testc.c:18]: Thanks for writing tests! See my other comment about the CI.

File: bindings/c/testc.c
Comment (elalish) [inline bindings/c/testc.c:77]: Should we make some `expect` calls to show when it fails? Does `GTest` have a C version?

File: bindings/c/manifoldc.cpp
Comment (elalish) [inline bindings/c/manifoldc.cpp:36]: Is there a way we can void this bool parameter by making `LevelSet` choose the `Seq` policy when the template parameter is not a Functor? Curious of @pca006132 knows any magic here.

File: bindings/c/manifoldc.cpp
Comment (elalish) [inline bindings/c/manifoldc.cpp:152]: It'd be nice to not have two functions here; see my other comment above. But I can't tell if there's a way around it.

**Review by elalish — CHANGES_REQUESTED** [review]
This is looking pretty good to me; mostly questions. What do you think, @pca006132?

File: src/polygon/CMakeLists.txt
Comment (geoffder) [reply src/polygon/CMakeLists.txt:18]: Since it is a sublibrary not used on it's own, this lets it compile separately still, but get linked into the main `a.`/`.so` being built. I was having some weird issues with relocating the many archives/objects when bundling for the ocaml library, and doing this to get rid of the extra library artifacts fixed it. The [entry](https://cmake.org/cmake/help/latest/command/add_library.html#object-libraries) in the cmake manual about it is most of what I know about OBJECT.

File: src/sdf/include/sdf.h
Comment (geoffder) [reply src/sdf/include/sdf.h:293]: Yeap! This means that dynamic lambdas can be used when CUDA is enabled without a runtime crash (something that the bindings can't avoid).

File: src/sdf/include/sdf.h
Comment (geoffder) [reply src/sdf/include/sdf.h:348]: The template trick avoids CUDA, but languages that have locks on their runtime (like ocaml and python) will not allow their closures to be run in parallel (which `LevelSet` will do by handing the sdf function over to thrust when the policy is parallel. So the policy parameter here was a way of allowing the caller to specify sequential computation. In the C bindings, the overloaded `LevelSet` with policy param is used to provide a `manifold_level_set_seq` alternative to the original one.

As far as the check for `!CudaEnabled()`, I figured that theoretically the caller could be using a static lambda (not `std::function`) from C++ while not having CUDA disabled, so this guard prevent them from setting `ParUnseq` as their policy.

File: bindings/c/testc.c
Comment (geoffder) [reply bindings/c/testc.c:77]: I'm not totally sure, I hadn't thought farther about the tests other than what I needed to do along the way to make sure things were working as I added them. :sweat_smile:

Happy to hear suggestions on how the testing should be done, laid out. I'm pretty unfamiliar with how it is usually done in C/C++.

File: bindings/c/meshio.cpp
Comment (geoffder) [reply bindings/c/meshio.cpp:1]: It doesn't do a lot and I know there isn't much in that file, but I figured that generally people are going to want to write their models into some formats, and `assimp` bindings don't necessarily exist already. Should spare some duplication of effort if people don't have to write their own off-ramps. Outputting to ascii stl or 3mf is quick enough to do on your own, but the more efficient binary formats that assimp can provide are nice to have. Speaking of, I haven't read more into it, but the default that is output right now for stl is ascii, would be nice to have the option to write binstl from MeshIO. 

Is there a particular reason why you want to leave mesh output as a just for testing convenience rather than providing something as part of `Manifold`?

File: bindings/c/manifoldc.cpp
Comment (geoffder) [reply bindings/c/manifoldc.cpp:36]: `ParSeq` is fine for `std::function` though generally, it is only a problem from particular host languages that might be using the C bindings, so I wanted to provide a way to lock in sequential usage of the closure without adding another separate named method to the API.

File: bindings/c/manifoldc.cpp
Comment (geoffder) [reply bindings/c/manifoldc.cpp:152]: I commented on the other one, but I think it is either this, or making the caller to provide a "boolean" (int) flag specifying that execution should be forced sequential. Unfortunately the factors that require avoiding parallel execution of the closure are external and not encoded into the type system as the distinction between static and dynamic lambdas are.

File: src/manifold/src/constructors.cpp
Comment (geoffder) [reply src/manifold/src/constructors.cpp:381]: Oh I don't think I meant to get rid of that, sorry. I'll add that back in now!

File: .github/workflows/manifold.yml
Comment (geoffder) [reply .github/workflows/manifold.yml:38]: Having them only turned on for one build (maybe one with Par and CUDA enabled) and not failing CI sounds good to me. I just flipped it on everywhere to make sure they were working.

File: bindings/c/box.cpp
Comment (geoffder) [reply bindings/c/box.cpp:39]: C doesn't really do bool

File: bindings/c/include/conv.hh
Comment (geoffder) [reply bindings/c/include/conv.hh:10]: I think they were auto added my my editor, I should be able to remove them.

File: bindings/c/box.cpp
Comment (geoffder) [reply bindings/c/box.cpp:14]: Yeap that's right

File: bindings/c/include/conv.hh
Comment (geoffder) [reply bindings/c/include/conv.hh:1]: No reason other than I thought C++ headers were supposed to have `.hh` extensions.

File: bindings/c/include/types.h
Comment (geoffder) [reply bindings/c/include/types.h:30]: Yes it isn't, there is a `cglm` counterpart library. This way there isn't any dependency on `glm` to be able to understand the structs (at least that was my assumption). Not sure if there is a better way to handle the passing of vector arrays back and forth, I might have taken an overly cautious route.

File: src/utilities/include/public.h
Comment (geoffder) [reply src/utilities/include/public.h]: Done!

File: bindings/c/include/conv.hh
Comment (geoffder) [reply bindings/c/include/conv.hh]: Renamed: https://github.com/elalish/manifold/pull/292/commits/1f77ef77f14fd76ace5d0608a683637fb557c0f2

File: src/manifold/src/constructors.cpp
Comment (geoffder) [reply src/manifold/src/constructors.cpp:381]: Added back in here: https://github.com/elalish/manifold/pull/292/commits/20a2fb5663a6ff935ee18299944a106abee2a42f

File: test/mesh_test.cpp
Comment (geoffder) [reply test/mesh_test.cpp]: Ok that did the trick, test is being skipped so the mac build is "passing". :+1:

File: bindings/c/include/conv.hh
Comment (geoffder) [reply bindings/c/include/conv.hh]: Went through and pruned a bunch of auto-included things. The C++ LSP running in my emacs must be doing it.

File: bindings/c/meshio.cpp
Comment (elalish) [reply bindings/c/meshio.cpp:1]: Yes, the reason is that formats are complex and lots of code to maintain and other projects are already better at it than mine. I need some boundaries, or else a hell of a lot more collaborators! 

I refuse to support STL though: STL cannot represent a manifold mesh in any kind of reliable way, so using this library and outputting STL is self-defeating. Pretty much any other format will do, 3MF being probably ideal. 

Anyway, if you want to maintain it, you can, but be warned: someone's going to want UVs and then atlasing and eventually you'll have to write Blender all over.

File: bindings/c/testc.c
Comment (elalish) [reply bindings/c/testc.c:77]: GTest is the framework we use for the rest of our C++ tests and it looks like it'll be fine for your C too: https://stackoverflow.com/questions/5335268/is-google-test-ok-for-testing-c-code

It's pretty simple, I don't think you'll have any trouble adapting it from our existing tests.

File: bindings/c/manifoldc.cpp
Comment (elalish) [reply bindings/c/manifoldc.cpp:25]: Ah, okay I see now.

File: bindings/c/manifoldc.cpp
Comment (elalish) [reply bindings/c/manifoldc.cpp:141]: Okay this makes sense now; can you add a comment on these bindings so people know they should really only link to one or the other depending on their language?

File: bindings/c/include/types.h
Comment (elalish) [reply bindings/c/include/types.h:30]: Up to you as the user. I find glm to be an excellent library that makes my life much easier. If you want to include cglm for these bindings that's fine, but I don't have a strong opinion.

File: bindings/c/meshio.cpp
Comment (geoffder) [reply bindings/c/meshio.cpp:1]: I didn't mean to suggest re-writing what assimp is providing and increase your responsibilities, just to do what is basically done now -- letting assimp do it. My decision to use MeshIO basically just that the interface with assimp is already being provided by it, so it is simple to provide it along with the C bindings. If assimp bindings (or some native/bound serialization library) exists in the language that the C bindings are being used from, then there is no problem, otherwise using Manifold will also mean coming up with those solutions on their own. Does that make sense?

File: bindings/c/manifoldc.cpp
Comment (geoffder) [reply bindings/c/manifoldc.cpp:141]: Yeap I can do that.

File: bindings/c/meshio.cpp
Comment (elalish) [reply bindings/c/meshio.cpp:1]: Maybe what I really mean is, `Manifold` can handle any kind of crazy vertex properties you want, but I'm afraid someone will just look at my crappy little MeshIO function that only outputs vertex colors and say, "oh I guess this library doesn't work on my meshes" and move on. There are just so many ways to deal with mesh data; I'm doing my best not to make any assumptions, but once you output to a file you've necessarily picked one.

File: bindings/c/include/types.h
Comment (geoffder) [reply bindings/c/include/types.h:30]: Okay. Since there isn't actually any number crunching happening and it's just a struct of floats I think this way works well enough then.

File: src/sdf/include/sdf.h
Comment (elalish) [reply src/sdf/include/sdf.h:348]: > As far as the check for `!CudaEnabled()`, I figured that theoretically the caller could be using a static lambda (not `std::function`) from C++ while not having CUDA disabled, so this guard prevent them from setting `ParUnseq` as their policy.

I'm pretty sure this particular logic is in `autoPolicy` already; would be great if you can look it over to verify.

Comment (elalish) [reply src/sdf/include/sdf.h:348]: Oh, I see it now, thanks.

File: bindings/c/manifoldc.cpp
Comment (geoffder) [reply bindings/c/manifoldc.cpp:141]: Added here: https://github.com/elalish/manifold/pull/292/commits/e25964bad67db377c3c193f787d020ccb62d1fb5

File: bindings/c/meshio.cpp
Comment (geoffder) [reply bindings/c/meshio.cpp:1]: I don't think you have to worry about that too much! If people want more, they can always do it themselves with the data available from `Mesh`, I just think it's a nice batteries included solution to have output functions readily available, even if they aren't state of the art.

File: bindings/c/testc.c
Comment (geoffder) [reply bindings/c/testc.c:77]: Ok I'll give that a shot soon. How much additional coverage beyond what I have in there do you think I should add? Right now it just has a couple basics, plus some of the trickier parts like the sdf.

**Comment by pca006132** [comment]
Sorry I am not able to review this thoroughly as I just caught the flu, but it seems that the bindings are nicely written and will not affect the main code (except the weird macos error), I think this looks good to me.

**Comment by geoffder** [comment]
Started up a test file for the C bindings: https://github.com/elalish/manifold/pull/292/commits/5dd830936867902e6519e5d5b6895554cdf64570

File: bindings/c/testc.c
Comment (elalish) [reply bindings/c/testc.c:77]: What you have looks good; again since you've signed up to maintain these C bindings, these tests are mostly to make your own life easier. You're welcome to add more as you see fit.

File: .github/workflows/manifold.yml
Comment (elalish) [reply .github/workflows/manifold.yml:38]: Okay, I think this is the only part left to do for this PR, yes? I'm happy to merge once this is done.

File: bindings/c/testc.c
Comment (geoffder) [reply bindings/c/testc.c:77]: Roger. I'll add some more of what I had in `testc.c` to the new `manifoldc_test.cpp` so I can get rid of that lone executable, then populate further as needed. 👍

File: .github/workflows/manifold.yml
Comment (geoffder) [reply .github/workflows/manifold.yml]: I've broken out the building/testing of the C bindings into a separate job so that it can be marked as not required easily in your branch protection rules.

File: src/utilities/include/public.h
Comment (geoffder) [reply src/utilities/include/public.h:263]: The discrete (renamed `numComponents`) is just the number of discrete mesh components. So the new name is more clear.

**Review by elalish — APPROVED** [review]
Thanks, this looks good to me; I look forward to seeing what you do with it! Progress has been a bit slow on #290, but it's getting there and there will be a couple of follow-ons to finish updating the API. Once all that lands, you'll probably want to look into updating these bindings accordingly.

**Comment by geoffder** [comment]
Ok sounds good. :pray: Thanks for providing the library and being open to this addition!

**Pull request merged**
